drv/sprot-api/src/lib.rs

```diff
@@ -0,0 +1,490 @@
+
+#![no_std]
+extern crate memoffset;
+
+use crc::{Crc, CRC_16_XMODEM};
+use derive_idol_err::IdolError;
+// Chained if let statements are almost here.
+use if_chain::if_chain;
+use userlib::{sys_send, FromPrimitive};
+use zerocopy::*;
+
+// TODO: Audit that each MsgError is used and has some reasonable action.
+// While a diverse set of error codes may be useful for debugging it
+// clutters code that just has to deal with the error.
+// then consider adding a function that translates an error code
+// into the desired action, e.g. InvalidCrc and FlowError should both
+// result in a retry on the SP side and an ErrorRsp on the RoT side.
+
+#[derive(Copy, Clone, FromPrimitive, PartialEq, Eq, IdolError)]
+pub enum MsgError {
+    /// Transfer size is outside of maximum and minimum lenghts for message type.
+    BadTransferSize = 1,
+
+    /// Server restarted
+    // ServerRestarted = 2,
+
+    /// CRC check failed.
+    InvalidCrc = 3,
+
+    /// FIFO overflow/underflow
+    FlowError = 4,
+
+    /// Unsupported protocol version
+    UnsupportedProtocol = 5,
+
+    /// Unknown message
+    BadMessageType = 6,
+
+    /// Transfer size is outside of maximum and minimum lenghts for message type.
+    BadMessageLength = 7,
+
+    /// Error from Spi
+    SpiServerError = 8,
+
+    /// Message is too large
+    Oversize = 9,
+
+    /// Tx buffer is unexpectedly not Idle.
+    TxNotIdle = 10,
+
+    CannotAssertCSn = 11,
+
+    RotNotReady = 12,
+
+    RspTimeout = 13,
+
+    BadResponse = 14,
+
+    RotBusy = 15,
+
+    /// Feature is not implemented
+    NotImplemented = 16,
+
+    /// An error code reserved for the SP was used by the Rot
+    NonRotError = 17,
+
+    /// A message with version = 0 was received unexpectedly.
+    EmptyMessage = 18,
+    /// Unknown Errors are mapped to 0xff
+    Unknown = 0xff,
+}
+
+/// Protocol version
+pub const VERSION_IGNORE: u8 = 0; // To be ignored
+pub const VERSION_1: u8 = 1; // Supported message format
+
+// SPI has no flow control, i.e. the RoT has no busy state indication
+// visible to the SP.
+// In future designs, we could consider adding a "ROT_READY" in addition
+// to ROT_IRQ or use an interconect that does have flow control.
+//
+// The RoT places 0xB2 (Busy -> "B" "Z" -> 0xB2) in its transmit FIFO when
+// it is not prepared to service SPI IO. If the SP ever clocks out
+// data before the RoT is ready, the SP will read 0xB2.
+// This code should never be seen on the SPI bus. If seen as the first
+// byte in a message or on the SPI bus with a logic analyzer, that needs
+// to be investigated and fixed.
+//
+// It may mean that the two parties are out of phase, some RoT task
+// is hampering the sprot task from meeting its realtime requirements,
+// or there is some other bug.
+pub const VERSION_BUSY: u8 = 0xB2;
+
+/// SPI Message types will allow for multiplexing and forward compatibility.
+#[derive(Copy, Clone, Eq, PartialEq, FromPrimitive, Unaligned, AsBytes)]
+#[repr(u8)]
+pub enum MsgType {
```

Comment:
Is this where I would add a message type for sending an RoT update from the SP? Put another way, we had talked about the RoT injecting a session key over SWD, is this where we would add that message type?

Reply:
Possibly. I envisioned all authenticated messages would go through sprockets communication originally. Although currently, messages between the SP and RoT are just proxied messages from the host. https://github.com/oxidecomputer/sprockets/blob/main/common/src/msgs.rs

I'm certainly open to using this message path for this, but I'd like to open the option of converting all this hand-written serialization code to just use hubpack instead.

Reply:
All converted to hubpack whenever practical. (1-byte payloads and the sink test messages are not hubpack).